274 add some automated model predictive current control to gem control

manjeetjha070 opened the pull request:
I have renamed the curred mpc example file to "pmsm_ccs_mpc_dq_current_control.ipynb" and created a new file for finite control set as "pmsm_fcs_mpc_dq_current_control.ipynb".

manjeetjha070 committed:
Renamed current example file to "pmsm_ccs_mpc_dq_current_control.ipynb" and created new file for finite current control

manjeetjha070 committed:
Renamed current example file to "pmsm_ccs_mpc_dq_current_control.ipynb" and created new file for finite current control

manjeetjha070 committed:
renamed existing ccs mpc and added new file for fcs(finte control set) mpc.

manjeetjha070 committed:
renamed existing ccs mpc and added new file for fcs(finte control set) mpc.

XyDrKRulof commented on examples/model_predictive_controllers/pmsm_ccs_mpc_dq_current_control.ipynb:
Would be nice to change the title of this file now to make clear that it is the CCS example. Thanks!

XyDrKRulof commented on examples/model_predictive_controllers/pmsm_fcs_mpc_dq_current_control.ipynb:
Very nice work already, thank you! :) Some comments:

1. Change the introductionary text and title to reflect that this is an FCS example
2. Remove Gekko (and other unused libraries) - all mentionings, import and installation
3. Implement more than one step prediction for MPC. Utilize self.ph_ for this
4. Create an FCS environment in the gem.make comment. Currently, it is still CCS

Thanks for your effort and in case of questions feel free to reach out for us

XyDrKRulof reviewed (changes requested):
Already very nice work. For improvements, see my comments.

manjeetjha070 committed:
Upadted the title of pmsm_ccs_mpc_dq_current_control.ipynb and implemented comments on pmsm_fcs_mpc_dq_current_control.ipynb file

manjeetjha070 committed:
Upadted the title of pmsm_ccs_mpc_dq_current_control.ipynb and implemented comments on pmsm_fcs_mpc_dq_current_control.ipynb file

manjeetjha070 commented:
I have implemented the following changes plz. review.

1. Changed the introductionary text and title to reflect that this is an FCS example and CCS example.
2. Removed Gekko (and other unused libraries).
3. Implemented generalized MPC. 

Pending point,

1. To create a FCS environment, I would need a guidance for it.

XyDrKRulof reviewed (changes requested):
Hey Manjeet,

thanks for your good work - the FCS - MPC algorithm with prediction horizon > 1 is already looking really good :) Here are some further comments that we discussed:

1. Replace the "max_depth" variable in the simulate_sequence function with self.ph_ (which, for readability, you could also rename to self.prediction_horizon)
2. The self.voltage_vectors variable seems to be a mix out of abc and alphabeta coordinates - please change this table to the one you get directly out of the environment:
`self.subactions = -np.power(-1,env.get_wrapper_attr('physical_system')._converter._subactions)`
3. Instead of saving the abc voltage sequence of the lowest cost prediction, save (and return) the indices of them as provided in 2. (e.g., if the best voltage vector is (-1,1,-1) and it is in position 3 return 3)
4. Update epsilon_el for the future predictions and average it for the abc -> dq conversion:
` v_dq = self._forward_transformation(np.array([v_a, v_b, v_c]), epsilon_el + 0.5*omega*tau)`
`epsilon_next = epsilon + omega*tau`

If you have any questions regarding any of these points feel free to contact us! :)

manjeetjha070 committed:
Refactored FCS-MPC implementation to align with environment-provided utilities and improve prediction logic

- Replaced manual abc → αβ → dq voltage transformations with environment's native method:
- Updated voltage vector table to use finite set directly from the environment via:

- Now storing and returning subaction **indices** instead of full abc voltage vectors for better integration and comparison
- Updated epsilon handling in prediction loop:
  - Epsilon is now incremented as
  - dq-transformation uses averaged value:
- Replaced hardcoded max_depth with  (renamed from  for clarity)

Previously, I was doing manual calculations: converting abc voltages to αβ and then dq using custom logic.
Now all transformations use the environment's own coordinate conversion function and finite action space, making the controller more accurate and modular.

Thanks again for the helpful feedback!

manjeetjha070 committed:
Refactored FCS-MPC implementation to align with environment-provided utilities and improve prediction logic

- Replaced manual abc → αβ → dq voltage transformations with environment's native method:
- Updated voltage vector table to use finite set directly from the environment via:

- Now storing and returning subaction **indices** instead of full abc voltage vectors for better integration and comparison
- Updated epsilon handling in prediction loop:
  - Epsilon is now incremented as
  - dq-transformation uses averaged value:
- Replaced hardcoded max_depth with  (renamed from  for clarity)

Previously, I was doing manual calculations: converting abc voltages to αβ and then dq using custom logic.
Now all transformations use the environment's own coordinate conversion function and finite action space, making the controller more accurate and modular.

Thanks again for the helpful feedback!

manjeetjha070 commented:
Refactored FCS-MPC implementation to align with environment-provided utilities and improve prediction logic

- Replaced manual abc → αβ → dq voltage transformations with environment's native method: `abc_to_dq_space`
- Updated voltage vector table to use finite set directly from the environment via:
  `self.subactions = -np.power(-1, env.get_wrapper_attr('physical_system')._converter._subactions)`
- Now storing and returning subaction **indices** instead of full abc voltage vectors for better integration and comparison
- Updated epsilon handling in prediction loop:
  - Epsilon is now incremented as `epsilon_next = epsilon + omega * tau`
  - dq-transformation uses averaged value: `epsilon + 0.5 * omega * tau`
- Replaced hardcoded max_depth with `self.prediction_horizon` (renamed from `self.ph_` for clarity)

Previously, I was doing manual calculations: converting abc voltages to αβ and then dq using custom logic.
Now all transformations use the environment's own coordinate conversion function and finite action space, making the controller more accurate and modular.

Thanks again for the helpful feedback!

manjeetjha070 committed:
added (self.u_abc_k1  = self.u_lim * self.subactions) for calculation of u_abc from subaction state

manjeetjha070 committed:
added (self.u_abc_k1  = self.u_lim * self.subactions) for calculation of u_abc from subaction state

manjeetjha070 committed:
Revised Description for FCS-MPC

manjeetjha070 committed:
Revised Description for FCS-MPC

manjeetjha070 commented:
I have revised the description for FCS-MPC, now its ready to get merge.

manjeetjha070 committed:
added mpc controller in gem_controllers,though testing is pending

manjeetjha070 committed:
added mpc controller in gem_controllers,though testing is pending

manjeetjha070 committed:
1)In gym_controller file edited such that MPC controller should not wrap around GymElectricMotorAdapter class. 2)updated the MPC controller file to process directly from environment,PMSM works.

manjeetjha070 committed:
1)In gym_controller file edited such that MPC controller should not wrap around GymElectricMotorAdapter class. 2)updated the MPC controller file to process directly from environment,PMSM works.

manjeetjha070 commented:
Now, the MPC controller calculates the motor-specific state vector and electrical jacobian once in the control method, then uses these for prediction in _simulate_sequence. This makes the controller compatible with different motor types and avoids redundant calculations during the prediction loop. Unsupported motor types will raise a clear error.

manjeetjha070 committed:
[200~fix: avoid double-counting of cross-coupling terms in MPC model

manjeetjha070 committed:
[200~fix: avoid double-counting of cross-coupling terms in MPC model

manjeetjha070 committed:
I have updated my MPC controller that takes motor dynamics from model_constants

manjeetjha070 committed:
I have updated my MPC controller that takes motor dynamics from model_constants

manjeetjha070 committed:
Added MPC current controller with delay compensation

manjeetjha070 committed:
Added MPC current controller with delay compensation

manjeetjha070 committed:
support weighted cost function (w_d, w_q) and forward args via GemController.make

- Added w_d and w_q parameters to MPCCurrentController
- Updated cost function to use weighted errors
- Modified GemController.make to forward extra kwargs into MPC

manjeetjha070 committed:
support weighted cost function (w_d, w_q) and forward args via GemController.make

- Added w_d and w_q parameters to MPCCurrentController
- Updated cost function to use weighted errors
- Modified GemController.make to forward extra kwargs into MPC

manjeetjha070 committed:
added configurable delay compensation feature

manjeetjha070 committed:
added configurable delay compensation feature

manjeetjha070 committed:
added DeadTimeProcessor from environment and removed Extrapolation of reference

manjeetjha070 committed:
added DeadTimeProcessor from environment and removed Extrapolation of reference

manjeetjha070 committed:
added comments more clearly and corrected the sequence of Id and Iq

manjeetjha070 committed:
added comments more clearly and corrected the sequence of Id and Iq

manjeetjha070 committed:
added document for finite MPC

manjeetjha070 committed:
added mpc in gem_control.rst

manjeetjha070 committed:
changed top level syntax

manjeetjha070 committed:
Force add image file

manjeetjha070 committed:
Force add image file

bhk11 committed:
Merge remote-tracking branch 'origin/nightly' into 274-add-some-automated-model-predictive-current-control-to-gem-control

# Conflicts:
#	examples/model_predictive_controllers/pmsm_mpc_dq_current_control.ipynb

manjeetjha070 committed:
Fix figure paths in mpc_current_controller.rst

manjeetjha070 committed:
docs: Add docstrings to MPC controller

bhk11 committed:
modified changelog and pyptoject toml to make a new realease after merge of the changes in this branch.

bhk11 reviewed (approved):
Works fine for me in the mentioned limits in the changelog
